angr/exploration_techniques/loop_seer.py

```diff
@@ -0,0 +1,140 @@
+import angr
+from . import ExplorationTechnique
+
+import logging
+from collections import defaultdict
+
+l = logging.getLogger("angr.exploration_techniques.loop_seer")
+
+class LoopSeer(ExplorationTechnique):
+    """
+    This exploration technique monitors exploration and maintains all
+    loop-related data (well, currently it is just the loop trip counts, but feel
+    free to add something else).
+    """
+
+    def __init__(self, cfg=None, functions=None, loops=None, bound=None, bound_reached=None, discard_stash='spinning'):
+        """
+        :param cfg          : Normalized CFG is required.
+        :param functions    : Function(s) containing the loop(s) to be analyzed.
+        :param loops        : Loop(s) to be analyzed.
+        :param bound        : Limit the number of iteration a loop may be executed.
+        :param bound_reached: If provided, should be a function that takes a SimulationManager and returns
+                              a SimulationManager. Will be called when loop execution reach the given bound.
+                              Default to moving states that exceed the loop limit to a discard stash.
+        :param discard_stash: Name of the stash containing states exceeding the loop limit.
+        """
+
+        super(LoopSeer, self).__init__()
+        self.cfg = cfg
+        self.functions = functions
+        self.bound = bound
+        self.bound_reached = bound_reached
+        self.discard_stash = discard_stash
+
+        self.loops = {}
+
+        if type(loops) == angr.analyses.loopfinder.Loop:
+            loops = [loops]
+        
+        if type(loops) in (list, tuple) and all(type(l) == angr.analyses.loopfinder.Loop for l in loops):
+            for loop in loops:
+                self.loops[loop.entry_edges[0][0].addr] = loop
+
+        elif loops is not None:
+            raise TypeError('What type of loop is it?')
+
+    def setup(self, simgr):
+        if self.cfg is None:
+            self.cfg = self.project.analyses.CFGFast(normalize=True)
+        elif not self.cfg.normalized:
+            l.warning("LoopSeer uses normalized CFG. Recomputing the CFG...") 
+            cfg.normalize()
+            self.cfg = cfg
+
+        if type(self.functions) == str:
+            func = [self.project.kb.functions.function(name=self.functions)]
```

Comment:
I would assume when a CFG is given, its KB should be used to retrieve information. You should use `self.cfg.kb` instead of the project-level KB.